packages/klighd-core/src/proxy-view/proxy-view.tsx

```diff
@@ -0,0 +1,742 @@
+
+        // Calculate size
+        const size = Math.min(canvasWidth, canvasHeight) * this.sizePercentage * 0.08;
+        if (size !== this.currSize) {
+            // Size of proxies has changed, cannot reuse previous renderings
+            this.clearRenderings();
+        }
+        this.currSize = size;
+
+        //// Calculate transformations ////
+        const transformedOffScreenNodes = filteredOffScreenNodes.map(node => ({
+            node: node,
+            transform: this.getTransform(node, size, canvasWidth, canvasHeight, scroll, zoom)
+        }));
+
+        //// Apply clustering ////
+        const clusteredNodes = this.applyClustering(transformedOffScreenNodes, size, canvasWidth, canvasHeight);
+
+        // Render the proxies
+        const res = [];
+        for (const { node, transform } of clusteredNodes) {
+            // Create a proxy
+            const vnode = this.createSingleProxy(node, transform, ctx);
+            if (vnode) {
+                res.push(vnode);
+            }
+        }
+
+        return res;
+    }
+
+    /** Returns an object containing lists of all off-screen and on-screen nodes in `currRoot`.
+     * Note that an off-screen node's children aren't included in the list, e.g. only outer-most off-screen nodes are returned. */
+    private getOffAndOnScreenNodes(currRoot: SKNode, ctx: SKGraphModelRenderer): { offScreenNodes: SKNode[], onScreenNodes: SKNode[] } {
+        const depthMap = ctx.depthMap;
+        const viewport = ctx.viewport;
+        if (!depthMap || !viewport) {
+            // Not yet initialized
+            return { offScreenNodes: [], onScreenNodes: [] };
+        }
+
+        // For each node check if it's off-screen
+        const offScreenNodes = [];
+        const onScreenNodes = [];
+        for (const node of currRoot.children as SKNode[]) {
+            const region = depthMap.getProvidingRegion(node, viewport, ctx.renderOptionsRegistry);
+            if (region && !depthMap.isInBounds(region, viewport)) {
+                // Node out of bounds
+                offScreenNodes.push(node);
+            } else {
+                // Node in bounds
+                onScreenNodes.push(node);
+
+                if (node.children.length > 0) {
+                    // Has children, recursively check them
+                    const offAndOnScreenNodes = this.getOffAndOnScreenNodes(node, ctx);
+                    offScreenNodes.push(...offAndOnScreenNodes.offScreenNodes);
+                    onScreenNodes.push(...offAndOnScreenNodes.onScreenNodes);
+                }
+            }
+        }
+
+        return { offScreenNodes: offScreenNodes, onScreenNodes: onScreenNodes };
+    }
+
+    /** Applies clustering to all `offScreenNodes` until there's no more overlap. Cluster-proxies are returned as VNodes. */
+    private applyClustering(offScreenNodes: { node: SKNode, transform: TransformAttributes }[],
```

Comment:
I think this clustering problem can be solved more efficiently using a sweep line approach, might be worth looking into

Reply:
Judging by what I've read so far, the approach can be fit to work for clustering but requires complex data structures to actually reach the O(n*log(n)) runtime. With what would be implementable without going too deep down that rabbit hole, I may be able to improve the average case - though worst case would still be O(n²). I'll look at that when I have some spare time. Thanks for the suggestion!